File: ramana/src/com/tm/sorting/Order.java
package com.tm.sorting;

import java.lang.reflect.Array;
import java.util.Comparator;

public class Order {
    public static <T> void bubbleSort(T[] array, Comparator<T> comparator) {
        boolean sorted = false;
        while (!sorted) {
            sorted = true;
            for (int i = 1; i < array.length; i++) {
                if (comparator.compare(array[i - 1], array[i]) > 0) {
                    swap(array, i - 1, i);
                    sorted = false;
                }
            }
        }
    }

    public static <T> void selectionSort(T[] array, Comparator<T> comparator) {
        for (int i = 0; i < array.length; i++) {
            int minIndex = i;
            for (int j = i + 1; j < array.length; j++)
                if (comparator.compare(array[minIndex], array[j]) > 0)
                    minIndex = j;
            swap(array, i, minIndex);
        }
    }

    public static <T> void insertionSort(T[] array, Comparator<T> comparator) {
        for (int i = 1; i < array.length; i++) {
            int sortedUpTo = i - 1;
            if (comparator.compare(array[sortedUpTo], array[i]) > 0) {
                for (int j = sortedUpTo; j >= 0; j--) {
                    if (comparator.compare(array[j], array[j + 1]) > 0)
                        swap(array, j + 1, j);
                    else
                        break;
                }
            }
        }
    }

    public static <T> void mergeSort(T[] array, Comparator<T> comparator) {
        System.arraycopy(mergeSort(array, 0, array.length - 1, comparator), 0, array, 0, array.length);
    }

    public static <T> T[] mergeSort(T[] array, int start, int end, Comparator<T> comparator) {
        if (start >= end) {
            T[] result = (T[]) Array.newInstance(array[0].getClass(), 1);
            result[0] = array[start];
            return result;
        }
        int mid = (start + end) / 2;
        return merge(mergeSort(array, start, mid, comparator), mergeSort(array, mid + 1, end, comparator), comparator);
    }

    public static <T> T[] merge(T[] a, T[] b, Comparator<T> comparator) {
        if (a.length == 0)
            return b;
        if (b.length == 0)
            return a;
        T[] result = (T[]) Array.newInstance(a[0].getClass(), a.length + b.length);
        int i = 0;
        int j = 0;
        int k = 0;
        while (true) {
            if (comparator.compare(a[i], b[j]) > 0) {
                result[k++] = b[j++];
                if (j == b.length) {
                    System.arraycopy(a, i, result, k, a.length - i);
                    return result;
                }
            } else {
                result[k++] = a[i++];
                if (i == a.length) {
                    System.arraycopy(b, j, result, k, b.length - j);
                    return result;
                }
            }
        }
    }

    public static <T> void heapSort(T[] array, Comparator<T> comparator) {
        buildMaxHeap(array, comparator);
        int hithertoUnsortedArrayLength = array.length;
        for (int i = 0; i < array.length; i++) {
            hithertoUnsortedArrayLength--;
            swap(array, 0, hithertoUnsortedArrayLength);
            maxHeapify(array, 0, comparator, hithertoUnsortedArrayLength);
        }
    }

    public static <T> void maxHeapify(T[] array, int i, Comparator<T> comparator, int len) {
        int l = left(i);
        int r = right(i);
        int smallest = i;
        if (l < len && comparator.compare(array[i], array[l]) < 0)
            smallest = l;
        if (r < len && comparator.compare(array[smallest], array[r]) < 0)
            smallest = r;
        if (i != smallest) {
            swap(array, i, smallest);
            maxHeapify(array, smallest, comparator, len);
        }
    }

    public static <T> void buildMaxHeap(T[] array, Comparator<T> comparator) {
        for (int i = array.length / 2 - 1; i >= 0; i--)
            maxHeapify(array, i, comparator, array.length);
    }

    private static int left(int i) {
        return 2 * i + 1;
    }

    private static int right(int i) {
        return 2 * i + 2;
    }

    public static <T> void quickSort(T[] array, Comparator<T> comparator) {
        quickSort(array, 0, array.length - 1, comparator);
    }

    public static <T> void quickSort(T[] array, int start, int end, Comparator<T> comparator) {
        if (start >= end)
            return;
        int p = partition(array, start, end, comparator);
        quickSort(array, start, p - 1, comparator);
        quickSort(array, p + 1, end, comparator);
    }

    public static <T> int partition(T[] array, int start, int end, Comparator<T> comparator) {
        int p = start - 1;
        for (int i = start; i < end; i++) {
            if (comparator.compare(array[i], array[end]) < 0) {
                p++;
                swap(array, p, i);
            }
        }
        p++;
        swap(array, p, end);
        return p;
    }

    private static <T> void swap(T[] array, int i, int j) {
        T temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

}

File: ramana/pythonFiles/order.py
def bubbleSort(array, compare):
  sorted = False
  l = len(array)
  while sorted == False:
    sorted = True
    for i in range(1, l):
      if compare(array[i-1], array[i]) > 0:
        sorted = False
        swap(array, i, i-1)

def selectionSort(array, compare):
  l = len(array)
  for i in range(l):
    minIndex = i
    for j in range(i+1, l):
      if compare(array[minIndex], array[j]) > 0:
        minIndex = j
    swap(array, minIndex, i)

def insertionSort(array, compare):
  l = len(array)
  for i in range(1, l):
    sortedUpTo = i-1
    if compare(array[i-1], array[i]) > 0:
      swap(array, i, sortedUpTo)
      for j in range(sortedUpTo, 0, -1):
        if compare(array[j-1], array[j]) <= 0:
          break
        else:
          swap(array, j, j-1)

def mergeSort(array, compare):
  b = mergeSortInternal(array, 0, len(array)-1, compare)
  del(array[:])
  array.extend(b)

def mergeSortInternal(array, start, end, compare):
  if start >= end:
    return [array[start]]
  mid = (start + end) / 2
  return merge(mergeSortInternal(array, start, mid, compare), mergeSortInternal(array, mid+1, end, compare), compare)

def merge(a, b, compare):
  la = len(a)
  lb = len(b)
  c = []
  i = 0
  j = 0
  while (True):
    if compare(a[i], b[j]) > 0:
      c.append(b[j])
      j = j + 1
      if j == lb:
        c.extend(a[i:])
        break
    else:
      c.append(a[i])
      i = i + 1
      if i == la:
        c.extend(b[j:])
        break
  return c

def heapSort(array, compare):
  buildHeap(array, compare)
  size = len(array)
  for i in range(size):
    size -= 1
    swap(array, 0, size)
    heapify(array, 0, compare, size)

def buildHeap(array, compare):
  size = len(array)
  for i in range(size / 2 - 1, -1, -1):
    heapify(array, i, compare, size)

def heapify(array, i, compare, size):
  l = left(i)
  r = right(i)
  smallest = i
  if l < size and compare(array[i], array[l]) < 0:
    smallest = l
  if r < size and compare(array[smallest], array[r]) < 0:
    smallest = r
  if i != smallest:
    swap(array, smallest, i)
    heapify(array, smallest, compare, size)

def left(i):
  return 2*i + 1

def right(i):
  return 2*i + 2

def quickSort(array, compare):
  quickSortInternal(array, 0, len(array)- 1, compare)

def quickSortInternal(array, start, end, compare):
  if start >= end:
    return
  p = partition(array, start, end, compare)
  quickSortInternal(array, start, p-1, compare)
  quickSortInternal(array, p+1, end, compare)

def partition(array, start, end, compare):
  p = start-1
  for i in range(start, end):
    if compare(array[i], array[end]) < 0:
      p += 1
      swap(array, i, p)
  p += 1
  swap(array, p, end)
  return p

def swap(array, i, j):
  temp = array[i]
  array[i] = array[j]
  array[j] = temp
